src/storage/mvcc/txn.rs

```diff
@@ -146,49 +146,63 @@ impl<S: Snapshot> MvccTxn<S> {
             Mutation::Insert((key, value)) => (key, Some(value), true),
         };
 
-        {
-            if !options.skip_constraint_check {
-                if let Some((commit, write)) = self.reader.seek_write(&key, u64::max_value())? {
-                    // Abort on writes after our start timestamp ...
-                    // If exists a commit version whose commit timestamp is larger than or equal to
-                    // current start timestamp, we should abort current prewrite, even if the commit
-                    // type is Rollback.
-                    if commit >= self.start_ts {
-                        MVCC_CONFLICT_COUNTER.prewrite_write_conflict.inc();
-                        return Err(Error::WriteConflict {
-                            start_ts: self.start_ts,
-                            conflict_start_ts: write.start_ts,
-                            conflict_commit_ts: commit,
-                            key: key.into_raw()?,
-                            primary: primary.to_vec(),
-                        });
-                    }
-                    if should_not_exist {
-                        if write.write_type == WriteType::Put
-                            || (write.write_type != WriteType::Delete
-                                && self.key_exist(&key, write.start_ts - 1)?)
-                        {
-                            return Err(Error::AlreadyExist {
-                                key: key.into_raw()?,
-                            });
-                        }
-                    }
-                }
+        if let Some(lock) = self.reader.load_lock(&key)? {
+            // Abort on lock belonging to other transaction.
+            if lock.ts != self.start_ts {
+                return Err(Error::KeyIsLocked {
+                    key: key.into_raw()?,
+                    primary: lock.primary,
+                    ts: lock.ts,
+                    ttl: lock.ttl,
+                });
+            }
+            if lock.lock_type != LockType::Pessimistic {
+                // Duplicated command. No need to overwrite the lock and data.
+                MVCC_DUPLICATE_CMD_COUNTER_VEC.prewrite.inc();
+                return Ok(());
             }
-            // ... or locks at any timestamp.
-            if let Some(lock) = self.reader.load_lock(&key)? {
-                if lock.ts != self.start_ts {
-                    return Err(Error::KeyIsLocked {
+        } else if options.prewrite_pessimistic_lock {
+            // Pessimistic lock does not exist, the transaction should be aborted.
```

Comment:
I think we needn't consider if the pessimistic lock already exist before priwrite? make the logical easier?

Reply:
em, I can't get your points. `options.prewrite_pessimistic_lock` means the transaction wants to prewrite a pessimistic lock. If the pessimistic lock doesn't exist, the transaction must abort because it may violate isolation.